### PR TITLE
gRPC Plugin Producer [WIP]

### PR DESCRIPTION
**Pull request by malnick**
This PR adds a [gRPC](http://www.grpc.io/docs/index.html) style plugin mechanism for registering third party plugins and sending them metrics aggregated by this service.

### CODE REVIEW

File: producers/plugin/plugin.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"golang.org/x/net/context"
```
Comment (darkonie) [inline producers/plugin/plugin.go]: since go1.7 this is called `"context"`

File: producers/plugin/plugin.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+		config:      cfg,
+		metricsChan: make(chan producers.MetricsMessage),
+	}
+	return &p, p.metricsChan
+}
+
+func (p *producerImpl) Run() error {
+	listener, err := net.Listen("tcp", fmt.Sprintf("%d", p.config.Port))
+	if err != nil {
+		plugLog.Fatalf("Failed to start TCP listening")
```
Comment (jdef) [inline producers/plugin/plugin.go]: generally not a fan of fatal logging outside of bootstrap/main funcs. it makes unit testing hard. usually better off returning the err and letting someone else higher up the chain decide whether to fatal or not

File: producers/plugin/plugin.go (hunk 1 above)
Comment (malnick) [inline producers/plugin/plugin.go]: The proto lib uses this context lib specifically, it's required.

File: config.go [hunk 3]
```diff
@@ -222,9 +227,9 @@ func getNewConfig(args []string) (Config, error) {
 	// Note: .getNodeInfo() is last so we are sure we have all the
 	// configuration we need from flags and config file to make
 	// this run correctly.
-	if err := c.getNodeInfo(); err != nil {
-		return c, err
-	}
+	//	if err := c.getNodeInfo(); err != nil {
+	//		return c, err
+	//	}
```
Comment (malnick) [inline config.go]: Uncomment this before merge.

File: producers/plugin/plugin.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	plugLog.Info("Starting plugin gRPC/TCP listening service")
+	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", p.config.Port))
+	if err != nil {
+		plugLog.Fatalf("Failed to start TCP listening, %v", err)
+	}
+	s := grpc.NewServer()
+	pb.RegisterMetricsServer(s, &metricsServerImpl{
+		metricsChan: p.metricsChan,
+	})
+	reflection.Register(s)
```
Comment (malnick) [inline producers/plugin/plugin.go]: Do not need gRPC reflect yet.

**Pull request closed without merging**